1.8.0 Gradle changes (#3333)
JetBrains/kotlin-web-site: docs/topics/gradle/gradle-configure-project.md

</tab>
</tabs>

When configuring your project, check the Kotlin Gradle plugin compatibility with available Gradle versions. 

> In the following table, there are minimum and maximum fully supported versions of Gradle and Android Gradle plugin:

Author: ```suggestion
In the following table, there are the minimum and maximum fully supported versions of Gradle and Android Gradle plugin:
```

> | Kotlin version | Gradle min and max versions              | Android Gradle plugin min and max versions            |

Reviewer: I am not sure about "max versions" - sounds like greater version will not work :thinking: 

Could we for now use "max fully supported versions"?

Author: This is written in the sentence just before the table:
>In the following table, there are minimum and maximum fully supported versions of Gradle and Android Gradle plugin:

I can mark fully supported versions with bold. Does it work?

Of course, I can add words "fully supported" to the table itself, but I avoided it for not making the table header too large.

Reviewer: > I can mark fully supported versions with bold. Does it work?

Nah, people will ignore it. Better then to add also following sentence:
> Latest Gradle and AGP versions should generally work without issues.

Author: Ok, I'll add. Now it looks so (tried bold). This is a preview from IDEA, on stagings %kotlinVersion% will be replaced with 1.8.0, but builds don't work right now😁
<img width="866" alt="Screenshot 2022-12-21 at 22 00 48" src="https://user-images.githubusercontent.com/78360457/209001975-37a5bf9a-6d10-4caa-9b19-e43b64ff3dfb.png">

|----------------|------------------------------------------|-------------------------------------------------------|

> | 1.8.0          | %minGradleVersion% – %maxGradleVersion%  | %minAndroidGradleVersion% – %maxAndroidGradleVersion% |   

Reviewer: will placeholder being replaced on actual merge?

Author: Sure, even not before the merge but soon, after a TeamCity successfully builds staging. The values are taken from here https://github.com/JetBrains/kotlin-web-site/pull/3333/files#diff-142a214d21f3012db045510fef72790293d95f0b2de8865cc16396783d723c0d

* `warning` – the Kotlin Gradle plugin will print a warning message; the default value for projects on Gradle less than 8.0.
* `ignore` – the plugin will skip the check and won't produce any messages.

To avoid JVM target incompatibility, [configure a toolchain](#gradle-java-toolchains-support) or align JVM versions manually.

#### What can go wrong if not checking targets compatibility {initial-collapse-state="collapsed"}

There are two ways of manually setting JVM targets for Kotlin and Java source sets:
* The implicit way via [setting up a Java toolchain](#gradle-java-toolchains-support).
* The explicit way via setting the `jvmTarget` attribute in the `kotlin` task and `targetCompatibility` 
  in the `java` extension or task.

If you use the second variant or have a default configuration with JDK not equal to `1.8`, different values of `jvmTarget` 
and `targetCompatibility` cause JVM target incompatibility. Let's consider a default configuration of JVM targets 
when you have only the Kotlin JVM plugin in your build script and no additional settings for JVM targets:

```kotlin
    plugins {

>         kotlin("jvm") version "%kotlinVersion%"

Author: Here is the same question about using the kotlinVersion variable https://github.com/JetBrains/kotlin-web-site/pull/3333/files#r1053512077

Configure the behavior of this check by setting the `kotlin.jvm.target.validation.mode` property in the `build.gradle`
file to:

* `error` – the plugin will fail the build; the default value for projects on Gradle 8.0+.
* `warning` – the Kotlin Gradle plugin will print a warning message; the default value for projects on Gradle less than 8.0.
* `ignore` – the plugin will skip the check and won't produce any messages.

To avoid JVM target incompatibility, [configure a toolchain](#gradle-java-toolchains-support) or align JVM versions manually.

#### What can go wrong if not checking targets compatibility {initial-collapse-state="collapsed"}

There are two ways of manually setting JVM targets for Kotlin and Java source sets:
* The implicit way via [setting up a Java toolchain](#gradle-java-toolchains-support).
* The explicit way via setting the `jvmTarget` attribute in the `kotlin` extension or task and `targetCompatibility` 
  in the `java` extension or task.

If you use the second variant or have a default configuration with JDK not equal to `1.8`, different values of `jvmTarget` 

> and `targetCompatibility` cause JVM target incompatibility. Let's consider a default configuration of JVM targets 

Reviewer: First sentence sounds weird to me. What does it try to say? :thinking:

Author: If you explicitly set the `jvmTarget` attribute in the `kotlin` extension or task and `targetCompatibility` 
  in the `java` extension or task
or have a default configuration with JDK not equal to `1.8`, different values of `jvmTarget` 
and `targetCompatibility` cause JVM target incompatibility.

The second variant is about the second item from the list above.

Reviewer: I don't know - still it is hard to read. WDYT about removing it and jump strait to example? At the end of example we showing what and why could go wrong.

>
{type="note"}

For related tasks like these, the Kotlin Gradle plugin checks for JVM target compatibility. Different values of 
the [`jvmTarget` attribute](gradle-compiler-options.md#attributes-specific-to-jvm) in the `kotlin` extension or task 
and [`targetCompatibility`](https://docs.gradle.org/current/userguide/java_plugin.html#sec:java-extension)
in the `java` extension or task cause JVM target incompatibility. For example:
the `compileKotlin` task has `jvmTarget=1.8`, and
the `compileJava` task has (or [inherits](https://docs.gradle.org/current/userguide/java_plugin.html#sec:java-extension)) `targetCompatibility=15`.

Configure the behavior of this check by setting the `kotlin.jvm.target.validation.mode` property in the `build.gradle`
file to:

> * `error` – the plugin will fail the build; the default value for projects on Gradle 8.0+.
> * `warning` – the Kotlin Gradle plugin will print a warning message; the default value for projects on Gradle less than 8.0.

Reviewer 2: ```suggestion
* `error` – the plugin fails the build; the default value for projects on Gradle 8.0+.
* `warning` – the plugin prints a warning message; the default value for projects on Gradle less than 8.0.
```

> * `ignore` – the plugin will skip the check and won't produce any messages.

Reviewer 2: ```suggestion
* `ignore` – the plugin skips the check and doesn't produce any messages.
```

To avoid JVM target incompatibility, [configure a toolchain](#gradle-java-toolchains-support) or align JVM versions manually.

> #### What can go wrong if not checking targets compatibility {initial-collapse-state="collapsed"}

Reviewer 2: ```suggestion
#### What can go wrong if targets aren't checked for compatibility {initial-collapse-state="collapsed"}
```

Author: Do you think we need to use passive here?
Maybe "What can go wrong if you don't set up checking targets compatibility"?

Author: @sarahhaggarty ?

There are two ways of manually setting JVM targets for Kotlin and Java source sets:
* The implicit way via [setting up a Java toolchain](#gradle-java-toolchains-support).
* The explicit way via setting the `jvmTarget` attribute in the `kotlin` extension or task and `targetCompatibility` 
  in the `java` extension or task.

> JVM target incompatibility occur if you:

Author: @Tapchicoma 
You said (this comment disappeared because I made a change) this part was hard to read and proposed deleting it and starting with the example. Well, I want to offer another variant: I tried to make the text more readable, [here](https://github.com/JetBrains/kotlin-web-site/pull/3333/commits/f90e40c026c6a5664a70432b4a07f9b1140b0088) you can see the change. I wouldn't like to start with an example, I always try to write some introductory sentences. Hope it works now; let me know if it's still hard to follow.

Reviewer: yeah - now looks better. Let's leave it as it is now

> JVM target incompatibility occur if you:

Reviewer 2: ```suggestion
JVM target incompatibility occurs if you:
```

kotlin.stdlib.default.dependency=false
```

#### Versions alignment of transitive dependencies

If you explicitly write the Kotlin version 1.8.0 or higher in your dependencies, for example: 
`implementation("org.jetbrains.kotlin:kotlin-stdlib:1.8.0")`, then the Kotlin Gradle Plugin uses this Kotlin version 
for transitive `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` dependencies. This is for avoiding class duplication from 

> different stdlib versions (learn more about merging `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` into `kotlin-stdlib`(whatsnew18.md#updated-jvm-compilation-target)). 

Author: ```suggestion
different stdlib versions ([learn more about merging `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` into `kotlin-stdlib`](whatsnew18.md#updated-jvm-compilation-target)). 
```

Author: Will do this right before merging.

You can disable this behavior with the `kotlin.stdlib.jdk.variants.version.alignment` Gradle property:

```properties
 `kotlin.stdlib.jdk.variants.version.alignment=false`
```

##### Other ways to align versions {initial-collapse-state="collapsed"}

* In case you don't use the Kotlin Gradle plugin or have issues with versions alignment, align all versions 
  via the Kotlin [BOM](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import). 
  Declare a platform dependency on `kotlin-bom` in your build script:

  ```kotlin
  implementation(platform("org.jetbrains.kotlin:kotlin-bom:%kotlinVersion%"))
  ```

* If you don't have a standard library explicitly: `kotlin.stdlib.default.dependency=false` in your `gradle.properties`,
  but one of your dependencies transitively brings some old Kotlin stdlib version, for example, `kotlin-stdlib-jdk7:1.7.20` and 

>   another dependency transitively brings `kotlin-stdlib:1.8+` – in this case, you can require `1.8.0` versions of these

Author: Will it be ok to replace it with a variable kotlinVersion, which later will have values `1.8.10`, `1.9.0`, and so on?

Reviewer: I think - yes

  transitive libraries:

  ```kotlin
  dependencies {
      constraints {
          add("implementation", "org.jetbrains.kotlin:kotlin-stdlib-jdk7") {
              version {
                  require("1.8.0")
              }
          }
          add("implementation", "org.jetbrains.kotlin:kotlin-stdlib-jdk8") {
              version {
                  require("1.8.0")
              }
          }
      }
  }
  ```
  

> * If you have a Kotlin version equal to `1.8.0`: `implementation("org.jetbrains.kotlin:kotlin-stdlib:1.8.0")` and 

Author: Here the same question about using the kotlinVersion variable https://github.com/JetBrains/kotlin-web-site/pull/3333/files#r1053512077

Reviewer: Should be fine as well

  an old version (less than `1.8.0`) of a Kotlin Gradle plugin – update the Kotlin Gradle plugin:

  ```kotlin
  // replace `<...>` with the plugin name
  plugins {
      kotlin("<...>") version "%kotlinVersion%"
  }
  ```

* If you have an explicit old version (less than `1.8.0`) of `kotlin-stdlib-jdk7`/`kotlin-stdlib-jdk8`, for example, 
  `implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:SOME_OLD_KOTLIN_VERSION")`, and a dependency that 

>   transitively brings `kotlin-stdlib:1.8+`, replace your `kotlin-stdlib-jdk*:SOME_OLD_KOTLIN_VERSION` with 

Author: ```suggestion
  transitively brings `kotlin-stdlib:1.8+`, [replace your `kotlin-stdlib-jdk*:SOME_OLD_KOTLIN_VERSION` with 
```

Author: Will do this right before merging.

>   `kotlin-stdlib-jdk*:%kotlinVersion%`(whatsnew18.md#updated-jvm-compilation-target) or [exclude](https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html#sec:excluding-transitive-deps) 

Author: ```suggestion
  `kotlin-stdlib-jdk*:%kotlinVersion%`](whatsnew18.md#updated-jvm-compilation-target) or [exclude](https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html#sec:excluding-transitive-deps) 
```

Author: Will do this right before merging.

different stdlib versions (learn more about merging `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` into `kotlin-stdlib`(whatsnew18.md#updated-jvm-compilation-target)). 
You can disable this behavior with the `kotlin.stdlib.jdk.variants.version.alignment` Gradle property:

```properties
 `kotlin.stdlib.jdk.variants.version.alignment=false`
```

##### Other ways to align versions {initial-collapse-state="collapsed"}

* In case you have issues with versions alignment, align all versions via the Kotlin [BOM](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import). 
  Declare a platform dependency on `kotlin-bom` in your build script:

  ```kotlin
  implementation(platform("org.jetbrains.kotlin:kotlin-bom:%kotlinVersion%"))
  ```

* If you don't have a standard library explicitly: `kotlin.stdlib.default.dependency=false` in your `gradle.properties`,
  but one of your dependencies transitively brings some old Kotlin stdlib version, for example, `kotlin-stdlib-jdk7:1.7.20` and 
  another dependency transitively brings `kotlin-stdlib:1.8+` – in this case, you can require `1.8.0` versions of these
  transitive libraries:

  ```kotlin
  dependencies {

>       constraints {

Author: Please help with Groovy syntax for this example.

Author: @Tapchicoma

Reviewer: Same for Groovy

              }
          }
          add("implementation", "org.jetbrains.kotlin:kotlin-stdlib-jdk8") {
              version {
                  require("1.8.0")
              }
          }
      }
  }
  ```
  
* If you have a Kotlin version equal to `1.8.0`: `implementation("org.jetbrains.kotlin:kotlin-stdlib:1.8.0")` and 
  an old version (less than `1.8.0`) of a Kotlin Gradle plugin – update the Kotlin Gradle plugin:

  ```kotlin
  // replace `<...>` with the plugin name
  plugins {
      kotlin("<...>") version "%kotlinVersion%"
  }
  ```

* If you have an explicit old version (less than `1.8.0`) of `kotlin-stdlib-jdk7`/`kotlin-stdlib-jdk8`, for example, 
  `implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:SOME_OLD_KOTLIN_VERSION")`, and a dependency that 
  transitively brings `kotlin-stdlib:1.8+`, replace your `kotlin-stdlib-jdk*:SOME_OLD_KOTLIN_VERSION` with 
  `kotlin-stdlib-jdk*:%kotlinVersion%`(whatsnew18.md#updated-jvm-compilation-target) or [exclude](https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html#sec:excluding-transitive-deps) 
  a transitive `kotlin-stdlib:1.8+` from the library that brings it:
 
  ```kotlin

>   dependencies {

Author: Please help with Groovy syntax for this example.

Reviewer: ```groovy
    implementation("com.example:lib:1.0") {
        exclude group: "org.jetbrains.kotlin", module: "kotlin-stdlib"
    }
 ```

You can disable this behavior with the `kotlin.stdlib.jdk.variants.version.alignment` Gradle property:

```properties
 `kotlin.stdlib.jdk.variants.version.alignment=false`
```

##### Other ways to align versions {initial-collapse-state="collapsed"}

* In case you have issues with versions alignment, align all versions via the Kotlin [BOM](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import). 
  Declare a platform dependency on `kotlin-bom` in your build script:

  <tabs group="build-script">
  <tab title="Kotlin" group-key="kotlin">

  ```kotlin
  implementation(platform("org.jetbrains.kotlin:kotlin-bom:%kotlinVersion%"))
  ```

  </tab>
  <tab title="Groovy" group-key="groovy">

  ```groovy
  implementation platform('org.jetbrains.kotlin:kotlin-bom:%kotlinVersion%')
  ```

  </tab>
  </tabs>

* If you don't have a standard library explicitly: `kotlin.stdlib.default.dependency=false` in your `gradle.properties`,
  but one of your dependencies transitively brings some old Kotlin stdlib version, for example, `kotlin-stdlib-jdk7:1.7.20` and 

>   another dependency transitively brings `kotlin-stdlib:1.8+` – in this case, you can require `1.8.0` versions of these

Author: ```suggestion
  another dependency transitively brings `kotlin-stdlib:1.8+` – in this case, you can require `%kotlinVersion%` versions of these
```

  transitive libraries:

  ```kotlin
  dependencies {
      constraints {
          add("implementation", "org.jetbrains.kotlin:kotlin-stdlib-jdk7") {
              version {

>                   require("1.8.0")

Author: ```suggestion
                  require("%kotlinVersion%")
```

              }
          }
          add("implementation", "org.jetbrains.kotlin:kotlin-stdlib-jdk8") {
              version {

>                   require("1.8.0")

Author: ```suggestion
                  require("%kotlinVersion%")
```

              }
          }
      }
  }
  ```
  

> * If you have a Kotlin version equal to `1.8.0`: `implementation("org.jetbrains.kotlin:kotlin-stdlib:1.8.0")` and 

Author: ```suggestion
* If you have a Kotlin version equal to `%kotlinVersion%`: `implementation("org.jetbrains.kotlin:kotlin-stdlib:%kotlinVersion%")` and 
```

kotlin.stdlib.default.dependency=false
```

#### Versions alignment of transitive dependencies

If you explicitly write the Kotlin version 1.8.0 or higher in your dependencies, for example: 
`implementation("org.jetbrains.kotlin:kotlin-stdlib:1.8.0")`, then the Kotlin Gradle Plugin uses this Kotlin version 
for transitive `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` dependencies. This is for avoiding class duplication from 

> different stdlib versions.[Learn more about [merging `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` into `kotlin-stdlib`](whatsnew18.md#updated-jvm-compilation-target). 

Reviewer 2: ```suggestion
different stdlib versions. [Learn more about [merging `kotlin-stdlib-jdk7` and `kotlin-stdlib-jdk8` into `kotlin-stdlib`](whatsnew18.md#updated-jvm-compilation-target).
```

              version {
                  require("%kotlinVersion%")
              }
          }
          add("implementation", "org.jetbrains.kotlin:kotlin-stdlib-jdk8") {
              version {
                  require("%kotlinVersion%")
              }
          }
      }
  }
  ```

  </tab>
  </tabs>
  
* If you have a Kotlin version equal to `%kotlinVersion%`: `implementation("org.jetbrains.kotlin:kotlin-stdlib:%kotlinVersion%")` and 
  an old version (less than `1.8.0`) of a Kotlin Gradle plugin – update the Kotlin Gradle plugin:

  
  <tabs group="build-script">
  <tab title="Kotlin" group-key="kotlin">

  ```kotlin
  // replace `<...>` with the plugin name
  plugins {
      kotlin("<...>") version "%kotlinVersion%"
  }
  ```

  </tab>
  <tab title="Groovy" group-key="groovy">

  ```groovy
  // replace `<...>` with the plugin name
  plugins {
      id "org.jetbrains.kotlin.<...>" version "%kotlinVersion%"
  }
  ```

  </tab>
  </tabs>

* If you have an explicit old version (less than `1.8.0`) of `kotlin-stdlib-jdk7`/`kotlin-stdlib-jdk8`, for example, 
  `implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:SOME_OLD_KOTLIN_VERSION")`, and a dependency that 

>   transitively brings `kotlin-stdlib:1.8+`, replace your `kotlin-stdlib-jdk<7/8>:SOME_OLD_KOTLIN_VERSION` with 

Author: ```suggestion
  transitively brings `kotlin-stdlib:1.8+`, [replace your `kotlin-stdlib-jdk<7/8>:SOME_OLD_KOTLIN_VERSION` with 
```

Author: Will do this right before merging.
